src/sources/kafka.rs

```diff
@@ -720,32 +1219,138 @@ fn create_consumer(config: &KafkaSourceConfig) -> crate::Result<StreamConsumer<C
     Ok(consumer)
 }
 
-#[derive(Default)]
-struct CustomContext {
+type TopicPartition = (String, i32);
+
+/// Status returned by partition consumer tasks, allowing the coordination task
+/// to differentiate between a consumer exiting normally (after receiving an end
+/// signal) and exiting when it reaches the end of a partition
+enum PartitionConsumerStatus {
+    NormalExit,
+    PartitionEOF,
+}
+
+enum KafkaCallback {
+    PartitionsAssigned(Vec<TopicPartition>, SyncSender<()>),
+    PartitionsRevoked(Vec<TopicPartition>, SyncSender<()>),
+    ShuttingDown(SyncSender<()>),
+}
+
+struct KafkaSourceContext {
+    acknowledgements: bool,
     stats: kafka::KafkaStatisticsContext,
-    finalizer: OnceLock<Arc<OrderedFinalizer<FinalizerEntry>>>,
+
+    /// A callback channel used to coordinate between the main consumer task and the acknowledgement task
+    callbacks: OnceLock<UnboundedSender<KafkaCallback>>,
+
+    /// A weak reference to the consumer, so that we can commit offsets during a rebalance operation
+    consumer: OnceLock<Weak<StreamConsumer<KafkaSourceContext>>>,
 }
 
-impl CustomContext {
-    fn new(expose_lag_metrics: bool) -> Self {
+impl KafkaSourceContext {
+    fn new(expose_lag_metrics: bool, acknowledgements: bool) -> Self {
         Self {
             stats: kafka::KafkaStatisticsContext { expose_lag_metrics },
-            ..Default::default()
+            acknowledgements,
+            callbacks: OnceLock::default(),
+            consumer: OnceLock::default(),
+        }
+    }
+
+    pub fn shutdown(&self) {
+        if let Some(tx) = self.callbacks.get() {
+            let (send, rendezvous) = sync_channel(0);
+            if tx.send(KafkaCallback::ShuttingDown(send)).is_ok() {
+                while rendezvous.recv().is_ok() {
+                    self.commit_consumer_state();
+                }
+            }
+        }
+    }
+
+    /// Emit a PartitionsAssigned callback with the topic-partitions to be consumed,
+    /// and block until confirmation is received that a stream and consumer for
+    /// each topic-partition has been set up. This function blocks until the
+    /// rendezvous channel sender is dropped by the callback handler.
+    fn consume_partitions(&self, tpl: &TopicPartitionList) {
+        let callbacks = self
+            .callbacks
+            .get()
+            .expect("Callbacks handler was not initialized.");
```

Comment:
If this is necessary (ref my question above), I recommend wrapping this pattern in an extra method to avoid the duplication.